File: src/lib/redis-scripts/llm-token-bucket.ts
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import type IORedis from 'ioredis';
import { createLogger } from '@/lib/logger';

const log = createLogger('redis-scripts:llm-token-bucket');

/**
 * Resolve `llm-token-bucket.lua` next to this module under both Bun and Node.
 * Same approach as `tenant-semaphore.ts`: `import.meta.url` works in both
 * runtimes, sidestepping the missing `__dirname` under native ESM.
 *
 * The script body is read at module load and cached; `defineCommand` ships
 * the source to Redis once per connection.
 */
const here = dirname(fileURLToPath(import.meta.url));
const SCRIPT = readFileSync(join(here, 'llm-token-bucket.lua'), 'utf8');

const COMMAND_NAME = 'llmTokenBucketAcquire';

/**
 * Define the Lua command on this client if not already defined.
 *
 * `defineCommand` attaches a method named `COMMAND_NAME` to the client
 * instance. We probe for that property to keep this call idempotent —
 * cheap on hot paths, defensive against double-registration warnings.
 */
function ensureCommand(redis: IORedis): void {
  // ioredis's dynamic-command typing isn't reflected in its `.d.ts`, so we
  // narrow through `unknown` to access the runtime-defined method.
  const dyn = redis as unknown as Record<string, unknown>;
  if (typeof dyn[COMMAND_NAME] === 'function') return;
  redis.defineCommand(COMMAND_NAME, { numberOfKeys: 2, lua: SCRIPT });
}

/**
 * Scope of a denial — which bucket refused the request.
 *
 * - `tenant`: the per-tenant bucket is empty (most common; means this user
 *   is sending too fast for their tier).
 * - `global`: the per-platform-key global bucket is empty (means *all*
 *   tenants combined are exceeding the upstream provider's rate limit).
 * - `config`: a configuration error — typically `cost > cap` — that can
 *   never succeed on retry. Callers should log and surface, not retry.
 */
export type DenyScope = 'tenant' | 'global' | 'config';

/**
 * Result of a token-bucket acquire attempt.
 *
 * Three variants, narrowed via `allowed` and `failedOpen`:
 *
 * 1. `{ allowed: true, tenantRemaining, globalRemaining, failedOpen?: false }`
 *    Happy path — call permitted, Redis updated. Remaining counts are real
 *    and safe to log / surface in `Retry-After`-style headers.
 *
 * 2. `{ allowed: true, failedOpen: true }` — Redis was unreachable; we let
 *    the call through (same trade-off as tenant-semaphore: better to serve
 *    traffic than wedge the worker pool during a Redis outage). Remaining
 *    counts are intentionally absent — we don't know the real values. Track
 *    `failedOpen` separately from real acquires in metrics; a spike means
 *    the limiter is effectively off and the upstream provider's 429s are
 *    the only guardrail.
 *
 * 3. `{ allowed: false, scope, retryMs }` — denied by one of the three
 *    scopes (`tenant` | `global` | `config`). See `DenyScope` for semantics.
 *
 * Narrowing pattern:
 * ```ts
 * const r = await tryAcquireLlmTokens(redis, opts);
 * if (!r.allowed) {
 *   // deny path — r.scope, r.retryMs
 * } else if (r.failedOpen) {
 *   // fail-open — no remaining counts available
 * } else {
 *   // happy path — r.tenantRemaining, r.globalRemaining accessible
 * }
 * ```
 */
export type AcquireLlmResult =
  | {
      allowed: true;
      tenantRemaining: number;
      globalRemaining: number;
      failedOpen?: false;
    }
  | {
      allowed: true;
      failedOpen: true;
    }
  | { allowed: false; scope: DenyScope; retryMs: number };

export interface AcquireLlmTokensOptions {
  /** Per-tenant bucket key, e.g. `llm:tenant:${userId}`. */
  tenantKey: string;
  /** Tenant bucket capacity (max burst). */
  tenantCap: number;
  /**
   * Tenant refill rate in tokens/sec. May be a float
   * (e.g. 1000 RPM → 16.6667/sec; 60 RPM → 1.0/sec).
   * Do NOT `Math.floor` this — pass the float through.
   */
  tenantRefillPerSec: number;
  /** Global bucket key, e.g. `llm:global:anthropic`. */
  globalKey: string;
  /** Global bucket capacity. */
  globalCap: number;
  /** Global refill rate in tokens/sec (float; same units as tenant). */
  globalRefillPerSec: number;
  /**
   * Cost of this acquire in tokens. Defaults to 1 (one request).
   * Pre-flight LLM-token estimation (LiteLLM-class) would pass an estimate
   * here, but that's a deferred follow-up — today we gate on requests.
   */
  cost?: number;
  /**
   * Override the timestamp passed to Lua. Tests use this for determinism;
   * production should let it default to `Date.now()`.
   */
  nowMs?: number;
}

// Lua reply shape:
//   allow: [1, tenantRemaining, globalRemaining]   (numbers)
//   deny:  [0, "tenant"|"global"|"config", retryMs] (string in slot 1)
type LuaReply = [number, number | string, number];

/**
 * Atomically check both the per-tenant and global Anthropic token buckets,
 * consuming one slot in each if both have capacity.
 *
 * Returns `{ allowed: true, ... }` when the call is permitted. When denied,
 * `scope` tells you which bucket refused and `retryMs` is a hint for the
 * earliest time the bucket will refill enough to admit `cost` tokens.
 *
 * Fails OPEN on Redis errors with `failedOpen: true` on the result. The
 * `log.warn` marks the occurrence for ops alerting.
 */
export async function tryAcquireLlmTokens(
  redis: IORedis,
  opts: AcquireLlmTokensOptions,
): Promise<AcquireLlmResult> {
  ensureCommand(redis);
  const cost = opts.cost ?? 1;
  const nowMs = opts.nowMs ?? Date.now();
  try {
    const dyn = redis as unknown as Record<
      string,
      (...args: unknown[]) => Promise<LuaReply>
    >;
    const raw = await dyn[COMMAND_NAME](
      opts.tenantKey,
      opts.globalKey,
      opts.tenantCap,
      opts.tenantRefillPerSec,
      opts.globalCap,
      opts.globalRefillPerSec,
      nowMs,
      cost,
    );
    const flag = Number(raw[0]);
    if (flag === 1) {
      return {
        allowed: true,
        tenantRemaining: Number(raw[1]),
        globalRemaining: Number(raw[2]),
      };
    }
    // Deny path: raw[1] is a scope string, raw[2] is retry_ms.
    const scope = String(raw[1]) as DenyScope;
    return {
      allowed: false,
      scope,
      retryMs: Number(raw[2]),
    };
  } catch (err) {
    log.warn(
      `llm-token-bucket acquire failed for ${opts.tenantKey} / ${opts.globalKey}, failing open: ${err instanceof Error ? err.message : String(err)}`,
    );
    // No remaining counts on the fail-open path — the type union excludes
    // them so callers must narrow on `failedOpen` before logging counts.
    return {
      allowed: true,
      failedOpen: true,
    };
  }
}

File: src/lib/redis-scripts/llm-token-bucket.lua
-- src/lib/redis-scripts/llm-token-bucket.lua
-- Atomic two-level token bucket — tenant bucket AND global bucket, gated in
-- one Lua call so a tenant can never be billed against the global budget
-- after the global already refused (and vice versa).
--
-- KEYS[1] = tenant bucket key,  e.g. "llm:tenant:user-abc"
-- KEYS[2] = global bucket key,  e.g. "llm:global:anthropic"
-- ARGV[1] = tenant cap          (integer)
-- ARGV[2] = tenant refill/sec   (float; e.g. 1000 RPM = 16.67/sec)
-- ARGV[3] = global cap          (integer)
-- ARGV[4] = global refill/sec   (float)
-- ARGV[5] = now (ms)            (integer; caller passes Date.now())
-- ARGV[6] = cost                (integer; usually 1 — one request)
--
-- Returns on allow: { 1, tenant_remaining, global_remaining }
-- Returns on deny:  { 0, "tenant"|"global"|"config", retry_ms }
--
-- Tokens are stored as FLOAT under the hash field 't' (e.g. 4.7). The
-- last-refill timestamp lives under 'ts' (ms). On global-deny after a
-- tenant-allow, we MUST refund the tenant token we just consumed — we use
-- HINCRBYFLOAT (HINCRBY would error because 't' is a float string) AND we
-- must restore the pre-refill `ts` so the next call doesn't double-credit
-- the refilled interval (the refund already restored tokens covering that
-- window; advancing `ts` would credit it a second time on the next acquire).
--
-- TTL of 3600s keeps abandoned tenants from polluting Redis forever.

local function refill(key, cap, rate_per_sec, now_ms, cost)
  local data = redis.call('HMGET', key, 't', 'ts')
  local tokens = tonumber(data[1])
  local last = tonumber(data[2])
  if tokens == nil or last == nil then
    -- Cold start: full bucket, last-refill is now.
    tokens = cap
    last = now_ms
  end
  -- Snapshot the pre-refill `ts` so the caller can restore it on a refund
  -- path (see global-deny branch in main). Without this, refunding tokens
  -- but leaving `ts = now_ms` lets the next acquire re-credit the same
  -- elapsed interval on top of the refund — double-credit.
  local prior_ts = last
  local elapsed_sec = (now_ms - last) / 1000.0
  if elapsed_sec < 0 then elapsed_sec = 0 end
  tokens = math.min(cap, tokens + elapsed_sec * rate_per_sec)
  if tokens < cost then
    local need = cost - tokens
    -- Guard against zero/negative refill rate (would otherwise divide-by-zero
    -- and return inf/nan). Treat as "never refills" → return a long retry.
    local retry_ms
    if rate_per_sec <= 0 then
      retry_ms = 60000
    else
      retry_ms = math.ceil((need / rate_per_sec) * 1000)
    end
    -- Persist the refilled-but-not-consumed token state so the next caller
    -- sees the up-to-date balance (and a stable ts for the elapsed math).
    redis.call('HMSET', key, 't', tokens, 'ts', now_ms)
    redis.call('EXPIRE', key, 3600)
    return {false, tokens, retry_ms, prior_ts}
  end
  tokens = tokens - cost
  redis.call('HMSET', key, 't', tokens, 'ts', now_ms)
  redis.call('EXPIRE', key, 3600)
  return {true, tokens, 0, prior_ts}
end

-- Defensive coercion: nil/non-numeric ARGV becomes a safe default that refuses
-- the call rather than crashing the script (which would fail-open in the TS
-- wrapper — silently bypassing the rate limiter).
local tenant_cap = tonumber(ARGV[1]) or 0
local tenant_rate = tonumber(ARGV[2]) or 0
local global_cap = tonumber(ARGV[3]) or 0
local global_rate = tonumber(ARGV[4]) or 0
local now_ms = tonumber(ARGV[5]) or 0
local cost = tonumber(ARGV[6]) or 1

-- Defensive guards: caps <= 0 mean "no budget configured / typo / disabled".
-- Refuse immediately under the appropriate scope rather than divide-by-zero
-- in the refill math.
if tenant_cap <= 0 then
  return {0, 'tenant', 0}
end
if global_cap <= 0 then
  return {0, 'global', 0}
end
-- A `cost` larger than the cap can never succeed — surface this as a
-- distinct scope so callers can log a config_error rather than retry forever.
if cost > tenant_cap or cost > global_cap then
  return {0, 'config', 0}
end

local t = refill(KEYS[1], tenant_cap, tenant_rate, now_ms, cost)
if not t[1] then
  return {0, 'tenant', t[3]}
end

local g = refill(KEYS[2], global_cap, global_rate, now_ms, cost)
if not g[1] then
  -- Refund the tenant slot we just took. Two writes are required:
  --   1. HINCRBYFLOAT restores the consumed token. HINCRBYFLOAT (not HINCRBY)
  --      because the hash field 't' holds a float string (e.g. "4.7");
  --      HINCRBY would error with "value is not an integer".
  --   2. HSET restores `ts` to the pre-refill value (`t[4]` = prior_ts).
  --      Otherwise the refilled-and-restored interval would be credited a
  --      second time on the next acquire (double-credit bug).
  redis.call('HINCRBYFLOAT', KEYS[1], 't', cost)
  redis.call('HSET', KEYS[1], 'ts', t[4])
  return {0, 'global', g[3]}
end

return {1, t[2], g[2]}
